EDMXTrimmer/EDMXTrimmer/EdmxTrimmer.cs

```diff
@@ -132,31 +132,48 @@ public void AnalyzeFile()
             List<XmlNode> entitySets, 
             List<XmlNode> entityTypes)
         {
-            string regex = EntitySearchTermsToRegularExpression(entitiesToKeep);
-            var entitiesKeep = entitySets.Where(n => Regex.IsMatch(n.Attributes[ATTRIBUTE_NAME].Value, regex)).ToList();
+            var (entitySetsToKeep, entityTypeNamesToKeep) = FilterByEntity(entitiesToKeep, entitySets, entityTypes, true);
 
-            RemoveEntitySets(entitySets, entitiesKeep);
-            RemoveEntityTypes(entityTypes, entitiesKeep);
+            RemoveEntitySets(entitySets, entitySetsToKeep);
+            RemoveEntityTypes(entityTypes, entityTypeNamesToKeep);
         }
 
         private void RemoveExcludedEntities(
             List<string> entitiesToExclude, 
             List<XmlNode> entitySets, 
             List<XmlNode> entityTypes)
         {
-            string regex = EntitySearchTermsToRegularExpression(entitiesToExclude);
-            var entitiesKeep = entitySets.Where(n => !Regex.IsMatch(n.Attributes[ATTRIBUTE_NAME].Value, regex)).ToList();
+            var (entitySetsToKeep, entityTypeNamesToKeep) = FilterByEntity(entitiesToExclude, entitySets, entityTypes, false);
 
-            RemoveEntitySets(entitySets, entitiesKeep);
-            RemoveEntityTypes(entityTypes, entitiesKeep);
+            RemoveEntitySets(entitySets, entitySetsToKeep);
+            RemoveEntityTypes(entityTypes, entityTypeNamesToKeep);
         }
 
-        private string EntitySearchTermsToRegularExpression(List<string> entitiesToKeep)
+        private (List<XmlNode> EntitySetsToKeep, IReadOnlyCollection<string> EntityTypeNamesToKeep) FilterByEntity(IEnumerable<string> filteringEntities, IEnumerable<XmlNode> entitySets, IEnumerable<XmlNode> entityTypes, bool includeFiltered) 
         {
-            List<string> listRegularExpression = entitiesToKeep.Select(s => EntitySearchTermToRegularExpression(s)).ToList();
-            string regex = String.Join("|", listRegularExpression.ToArray());
+            var nameRegex = EntitySearchTermsToRegularExpression(filteringEntities);
+
+            var entitySetsNodes = entitySets
+                .Where(n => Regex.IsMatch(n.Attributes[ATTRIBUTE_NAME].Value, nameRegex) ? includeFiltered : !includeFiltered)
+                .ToList();
+
+            var entityTypeNames = entitySetsNodes
+                .Select(n => GetEntityTypeWithoutNamespace(n, TAG_ENTITY_TYPE))
+                .Concat(entityTypes
+                    .Where(node => Regex.IsMatch(node.Attributes[ATTRIBUTE_NAME].Value, nameRegex) ? includeFiltered : !includeFiltered)
```

Comment:
:wrench:

Thanks again for the contributions. I had a chance to look at the commits of the smaller pr in detail. Looks pretty good so far.

This part needs some work, I think. I did some regression testing with D365FO edmx files and noticed that with the `--entitiestoexclude` option, the entities remained in the resulting file.

This seems to be caused by this change in how the `entityTypeNames` are determined. Previously, this was based only on the filtered entity sets (see below for the original logic). Now, it also adds the names from the directly filtered entity types. However, the entity type names are in singular, while the entity set name is in plural. The entity names for the include/exclude parameters so far have been given in plural. For the new version to work, they probably would need to be given both in plural and in singular. However, I'm not sure if that would work as filter for both entity sets and types.

Since the intent of this change is to enable filtering of entities that are not part of sets, one option I can think of is to provide additional parameters where the singular entity names are specified. Those would then only be applied to entitiy types that are not part of an entity set. Another simpler one might be to provide a `--excludeentitiesnotinsets` parameter to remove entities that are not part of sets alltogether.

https://github.com/shashisadasivan/EDMXTrimmer/blob/67957002f9d3831d2e4185039c87861e4f116849/EDMXTrimmer/EDMXTrimmer/EdmxTrimmer.cs#L184-L190